### DISCUSSION TITLE
Using ffi for event loop alternative to stream_select?

### DISCUSSION THREAD

**Discussion by plamenbotev** [reactphp/reactphp#533] [General]
Hello,

I was wondering if anyone tried using FFI for creating an event loop that could potentially be used if additional modules are not available and perhaps can be used OOTB and to avoid the limitations imposed by using stream_select.
I was looking at https://github.com/symplely/uv-ffi and was wondering if there would be significant differences of implementing an event loop like that instead of using a separate extension.

**Comment by WyriHaximus** [comment]
Think @clue tried it out a while ago but nothing solid came out of that back then. If a FFI package would expose the exact same API as the extension it could be a drop in replacement in those situations.

> was wondering if there would be significant differences of implementing an event loop like that instead of using a separate extension.

Not really, it's a matter of calling different functions achieving the same thing. It differs from event loop to event loop anyway.

**Comment by plamenbotev** [comment]
I was able to use the mentioned package and it worked OOTB.
I see difference in the requests per second using apache bench, but have not checked what exactly is the cause for the difference yet.

**Reply by WyriHaximus** [reply]
What was the different in requests per second, out of interest?

**Reply by plamenbotev** [reply]
My initial benchmark showed around 5000 req / sec when using the extension and around 3000 req / sec, when using the ffi bundle.

**Reply by SimonFrings** [reply]
> I was looking at https://github.com/symplely/uv-ffi and was wondering if there would be significant differences of implementing an event loop like that instead of using a separate extension.

Interesting, so one difference definitely seems to be performance, have you figured out the reason for this yet? I can see this being useful to avoid certain limitations, but I'm not quite convinced how much is really gained by this (depending on your use case of course) 🤔

**Reply by plamenbotev** [reply]
I have not investigated further as this was just a side project for experiments. My benchmark was performed on my laptop with stock installed PHP  without any optimisations, so definitely far from prod like environment . I have also not done benchmark using the loop based on stream_select.
I will have to repeat my tests and try to thinker the setup to see if there is any difference.
I suspect that there will be some overhead caused by the ffi. :)
